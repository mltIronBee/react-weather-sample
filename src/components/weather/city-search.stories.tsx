import React, { ComponentProps } from "react";
import { Story } from "@storybook/react/types-6-0";
import { CitySearch } from "@components/weather";

const Template: Story<ComponentProps<typeof CitySearch>> = (args) => <CitySearch {...args} />;

export const Default = Template.bind({});

Default.args = {
	value: "Odesa",
};

export const Loading = Template.bind({});

Loading.args = {
	value: "Odesa",
	searching: true,
};

export const WithError = Template.bind({});

WithError.args = {
	value: "Lorem Ipsum",
	errorMessage: "Search error",
};

export default {
	title: "CitySearch",
	component: CitySearch,
};
