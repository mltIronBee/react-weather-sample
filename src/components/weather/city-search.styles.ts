import { Theme } from "@material-ui/core";
import { makeStyles, createStyles } from "@material-ui/styles";

export default makeStyles((theme: Theme) =>
	createStyles({
		loadingIcon: {
			padding: theme.spacing() * 1.5,
		},
	}),
);
