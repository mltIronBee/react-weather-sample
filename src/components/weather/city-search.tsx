import React from "react";
import TextField from "@material-ui/core/TextField";
import InputAdornment from "@material-ui/core/InputAdornment";
import IconButton from "@material-ui/core/IconButton";
import Tooltip from "@material-ui/core/Tooltip";
import SearchIcon from "@material-ui/icons/Search";
import GpsIcon from "@material-ui/icons/GpsFixed";
import GpsOffIcon from "@material-ui/icons/GpsOff";
import CircularProgress from "@material-ui/core/CircularProgress";
import useStyles from "@components/weather/city-search.styles";
import { useTranslation } from "react-i18next";

export interface ICitySearchProps {
	value: string;
	searching: boolean;
	geolocationAvailable: boolean;
	errorMessage?: string;
	onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
	onSearch: (event: React.FormEvent<HTMLFormElement>) => void | Promise<void>;
	onGetLocation: () => void | Promise<void>;
}

export const CitySearch: React.FC<ICitySearchProps> = (props) => {
	const classes = useStyles();
	const { t } = useTranslation(["search-input"]);

	return (
		<form onSubmit={props.onSearch}>
			<TextField
				fullWidth
				id="city-search"
				label={t("search-input:label")} //"Search city"
				placeholder={t("search-input:placeholder")} //"Enter city to search"
				error={!!props.errorMessage}
				helperText={props.errorMessage}
				value={props.value}
				onChange={props.onChange}
				variant="outlined"
				InputProps={{
					endAdornment: (
						<InputAdornment position="end">
							<Tooltip
								title={
									props.geolocationAvailable
										? (t("search-input:use-location-tooltip") as string) /*"Use current location"*/
										: (t("search-input:location-not-available") as string) /*"Geolocation is not available"*/
								}
							>
								<span>
									<IconButton
										disabled={!props.geolocationAvailable}
										aria-label={
											props.geolocationAvailable
												? t("search-input:use-location-aria-label") /*"get current location"*/
												: t("search-input:location-not-available-aria")
										}
										onClick={props.onGetLocation}
										className={classes.iconButton}
									>
										{props.geolocationAvailable ? <GpsIcon /> : <GpsOffIcon />}
									</IconButton>
								</span>
							</Tooltip>
							{props.searching ? (
								<CircularProgress className={classes.loadingIcon} size={24} />
							) : (
								<Tooltip title={t("search-input:search-tooltip") as string}>
									<span>
										<IconButton
											disabled={!props.value}
											className={classes.iconButton}
											component="button"
											type="submit"
											aria-label={t("search-input:search-aria-label")}
										>
											<SearchIcon />
										</IconButton>
									</span>
								</Tooltip>
							)}
						</InputAdornment>
					),
				}}
			/>
		</form>
	);
};
