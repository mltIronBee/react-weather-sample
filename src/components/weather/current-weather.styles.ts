import { makeStyles } from "@material-ui/core/styles";

export default makeStyles({
	sunriseSunsetContainer: {
		width: "100%",
		display: "flex",
		justifyContent: "space-evenly",
	},
	sunriseIcon: {
		color: "#EC6E4C",
	},
	sunsetIcon: {
		color: "#EC6E4C",
	},
	weatherContainer: {
		display: "flex",
		justifyContent: "center",
		alignItems: "center",
	},
	weatherDescription: {
		display: "inline-block",
		textTransform: "capitalize",
	},
	weatherIcon: {
		display: "inline-block",
	},
	windDirectionIcon: (windDeg: number) => ({
		transform: `rotate(${windDeg}deg)`,
	}),
});
