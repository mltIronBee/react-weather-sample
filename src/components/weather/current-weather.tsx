import React, { memo, useMemo } from "react";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import Divider from "@material-ui/core/Divider";
import useStyles from "@components/weather/current-weather.styles";
import ListItemIcon from "@material-ui/core/ListItemIcon";
import DownIcon from "@material-ui/icons/VerticalAlignBottom";
import CloudIcon from "@material-ui/icons/Cloud";
import DropIcon from "@material-ui/icons/Opacity";
import SunIcon from "@material-ui/icons/WbSunny";
import SnowIcon from "@material-ui/icons/AcUnit";
import VisibilityIcon from "@material-ui/icons/Visibility";
import SpeedIcon from "@material-ui/icons/Speed";
import ArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import RainIcon from "@components/icons/rain";
import SunriseIcon from "@components/icons/sunrise";
import SunsetIcon from "@components/icons/sunset";
import { invertDirection } from "@utils/angle";
import { useTranslation } from "react-i18next";

export interface ICurrentWeatherProps {
	date: string;
	sunrise: string;
	sunset: string;
	temperature: number;
	feelsLike: number;
	pressure: number;
	humidity: number;
	clouds: number;
	uvIndex: number;
	visibility: number;
	rain?: number;
	snow?: number;
	windSpeed: number;
	windDeg: number;
	weatherIcon: string;
	weatherDescription: string;
}

const formatTemperature = (temperature: number): string =>
	temperature > 0 ? `+${Math.round(temperature)}` : Math.round(temperature).toString(10);

const getReadableWindDirection = (windDeg: number): string => {
	if (windDeg > 337.5 || windDeg <= 22.5) {
		return "current-weather:north";
	} else if (windDeg > 22.5 && windDeg <= 67.5) {
		return "current-weather:north-east";
	} else if (windDeg > 67.5 && windDeg <= 112.5) {
		return "current-weather:east";
	} else if (windDeg > 112.5 && windDeg <= 157.5) {
		return "current-weather:south-east";
	} else if (windDeg > 157.5 && windDeg <= 202.5) {
		return "current-weather:south";
	} else if (windDeg > 202.5 && windDeg <= 247.5) {
		return "current-weather:south-west";
	} else if (windDeg > 247.5 && windDeg <= 292.5) {
		return "current-weather:west";
	}

	return "current-weather:north-west";
};

export const CurrentWeather: React.FC<ICurrentWeatherProps> = memo((props) => {
	const invertedAngle = useMemo(() => invertDirection(props.windDeg), [props.windDeg]);
	const classes = useStyles(invertedAngle);
	const { t } = useTranslation(["current-weather", "units"]);

	return (
		<Grid container>
			<Grid item xs={12} md={5} lg={4} xl={3}>
				<Grid container direction="column" justify="center" alignItems="center">
					<Typography>{props.date}</Typography>
					<span className={classes.weatherContainer}>
						<img className={classes.weatherIcon} alt={props.weatherDescription} src={props.weatherIcon} />
						<Typography className={classes.weatherDescription} variant="h4">
							{props.weatherDescription}
						</Typography>
					</span>
					<Typography gutterBottom variant="h2">
						{formatTemperature(props.temperature)}&deg;
					</Typography>
					<Typography gutterBottom variant="h4">
						{t("current-weather:feels-like")}: {formatTemperature(props.feelsLike)}&deg;
					</Typography>
					<Typography gutterBottom component="span" className={classes.sunriseSunsetContainer}>
						<Typography align="center" variant="h5" display="inline">
							<SunriseIcon className={classes.sunriseIcon} /> {t("current-weather:sunrise-at")} {props.sunrise}
						</Typography>
						<Typography align="center" variant="h5" display="inline">
							<SunsetIcon className={classes.sunsetIcon} /> {t("current-weather:sunset-at")} {props.sunset}
						</Typography>
					</Typography>
				</Grid>
			</Grid>
			<Grid item xs={12} md={7} lg={8} xl={9}>
				<List>
					<ListItem>
						<ListItemIcon>
							<DownIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.pressure} hPa`} secondary={t("current-weather:atmospheric-pressure")} />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<DropIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.humidity} %`} secondary={t("current-weather:humidity")} />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<CloudIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.clouds} %`} secondary={t("current-weather:clouds")} />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<SunIcon />
						</ListItemIcon>
						<ListItemText primary={props.uvIndex} secondary={t("current-weather:uv-index")} />
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<VisibilityIcon />
						</ListItemIcon>
						<ListItemText primary={`${props.visibility} ${t("units:m")}`} secondary={t("current-weather:visibility")} />
					</ListItem>
					<Divider />
					{props.rain && (
						<React.Fragment>
							<ListItem>
								<ListItemIcon>
									<RainIcon />
								</ListItemIcon>
								<ListItemText primary={`${props.rain} ${t("units:mm")}`} secondary={t("current-weather:rain")} />
							</ListItem>
							<Divider />
						</React.Fragment>
					)}
					{props.snow && (
						<React.Fragment>
							<ListItem>
								<ListItemIcon>
									<SnowIcon />
								</ListItemIcon>
								<ListItemText primary={`${props.snow} ${t("units:mm")}`} secondary={t("current-weather:snow")} />
							</ListItem>
							<Divider />
						</React.Fragment>
					)}
					<ListItem>
						<ListItemIcon>
							<SpeedIcon />
						</ListItemIcon>
						<ListItemText
							primary={`${props.windSpeed} ${t("units:mps")}`}
							secondary={t("current-weather:wind-speed")}
						/>
					</ListItem>
					<Divider />
					<ListItem>
						<ListItemIcon>
							<ArrowUpwardIcon className={classes.windDirectionIcon} />
						</ListItemIcon>
						<ListItemText
							primary={t(getReadableWindDirection(invertedAngle))}
							secondary={t("current-weather:wind-direction")}
						/>
					</ListItem>
				</List>
			</Grid>
		</Grid>
	);
});

CurrentWeather.displayName = "CurrentWeather";
