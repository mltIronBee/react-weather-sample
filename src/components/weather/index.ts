export * from "@components/weather/city-search";
export * from "@components/weather/forecast-graph";
export * from "@components/weather/tab-controls";
export * from "@components/weather/current-weather";
