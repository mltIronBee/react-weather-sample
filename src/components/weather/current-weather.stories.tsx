import React, { ComponentProps, useState } from "react";
import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import Slider from "@material-ui/core/Slider";
import { DateTime } from "luxon";
import { Story } from "@storybook/react/types-6-0";
import { CurrentWeather, ICurrentWeatherProps } from "@components/weather";
import { getWeatherIcon } from "@utils/weather-api";

const WindDirectionWrapper: React.FC<Omit<ICurrentWeatherProps, "windDeg">> = (props) => {
	const [windDeg, setWindDeg] = useState(0);

	const handleWindDirectionChange = (event: React.ChangeEvent<unknown>, value: number | number[]): void => {
		setWindDeg(value as number);
	};

	return (
		<Grid container>
			<Typography gutterBottom>Wind direction</Typography>
			<Grid spacing={3} container>
				<Grid item>
					<Typography>0</Typography>
				</Grid>
				<Grid item xs>
					<Slider min={0} max={360} value={windDeg} onChange={handleWindDirectionChange} />
				</Grid>
				<Grid item>
					<Typography>360</Typography>
				</Grid>
			</Grid>
			<Grid container>
				<CurrentWeather windDeg={windDeg} {...props} />
			</Grid>
		</Grid>
	);
};

const Template: Story<ComponentProps<typeof WindDirectionWrapper>> = (args) => <WindDirectionWrapper {...args} />;

export const Default = Template.bind({});

Default.args = {
	date: DateTime.local().toLocaleString(DateTime.DATETIME_FULL),
	sunrise: "06:00",
	sunset: "18:00",
	temperature: 24,
	feelsLike: 20,
	pressure: 1000,
	humidity: 80,
	clouds: 0,
	uvIndex: 0,
	visibility: 10000,
	windSpeed: 5,
	weatherIcon: getWeatherIcon("01d"),
	weatherDescription: "Clear sky",
};

export const WithRain = Template.bind({});

WithRain.args = {
	date: DateTime.local().toLocaleString(DateTime.DATETIME_FULL),
	sunrise: "06:00",
	sunset: "18:00",
	temperature: 5,
	feelsLike: -3,
	pressure: 1000,
	humidity: 80,
	clouds: 0,
	uvIndex: 0,
	visibility: 10000,
	windSpeed: 5,
	weatherIcon: getWeatherIcon("09d"),
	weatherDescription: "shower rain",
	rain: 30,
};

export const WithSnow = Template.bind({});

WithSnow.args = {
	date: DateTime.local().toLocaleString(DateTime.DATETIME_FULL),
	sunrise: "06:00",
	sunset: "18:00",
	temperature: -14,
	feelsLike: -20,
	pressure: 1000,
	humidity: 80,
	clouds: 0,
	uvIndex: 0,
	visibility: 10000,
	windSpeed: 5,
	weatherIcon: getWeatherIcon("13d"),
	weatherDescription: "Snow",
	snow: 30,
};

export default {
	title: "Current weather component",
	component: CurrentWeather,
};
