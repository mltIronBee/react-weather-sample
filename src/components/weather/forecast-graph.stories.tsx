import React, { ComponentProps } from "react";
import { Story } from "@storybook/react/types-6-0";
import { ForecastGraph } from "@components/weather";
import { IForecastGraphProps } from "@components/weather/forecast-graph";

const sampleData: IForecastGraphProps["data"] = [
	{ dayOfWeek: 1, minTemperature: 5, maxTemperature: 13 },
	{ dayOfWeek: 2, minTemperature: 3, maxTemperature: 11 },
	{ dayOfWeek: 3, minTemperature: 4, maxTemperature: 9 },
	{ dayOfWeek: 4, minTemperature: 4, maxTemperature: 9 },
	{ dayOfWeek: 5, minTemperature: 4, maxTemperature: 9 },
	{ dayOfWeek: 6, minTemperature: 2, maxTemperature: 8 },
	{ dayOfWeek: 7, minTemperature: 1, maxTemperature: 8 },
];

const Template: Story<ComponentProps<typeof ForecastGraph>> = ({ data, ...args }) => (
	<ForecastGraph data={data || sampleData} {...args} />
);

export const Default = Template.bind({});

export const WithoutData = Template.bind({});

WithoutData.args = {
	data: [],
};

export const WithError = Template.bind({});

WithError.args = {
	hasError: true,
};

export const LoadingWithoutData = Template.bind({});

LoadingWithoutData.args = {
	data: [],
	loading: true,
};

export const LoadingWithData = Template.bind({});

LoadingWithData.args = {
	loading: true,
};

export const LoadingWithError = Template.bind({});

LoadingWithError.args = {
	loading: true,
	hasError: true,
};

export default {
	title: "Forecast chart component",
	component: ForecastGraph,
	decorators: [
		(Story: React.ComponentType): JSX.Element => (
			<div style={{ width: "50%", height: 400 }}>
				<Story />
			</div>
		),
	],
};
