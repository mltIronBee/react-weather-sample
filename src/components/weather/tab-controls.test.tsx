import React from "react";
import { render } from "@utils/test-utils";
import { TabControls } from "@components/weather";
import "@testing-library/jest-dom/extend-expect";

describe("Tab controls component", () => {
	describe("Rendering", () => {
		const voidCallback = (): void => undefined;

		it("Should render correctly on large screen", () => {
			const { container } = render(
				<TabControls currentWeatherDisabled={false} largeScreen onTabChange={voidCallback} currentTab={0} />,
			);

			expect(container.firstChild).toMatchSnapshot();
		});

		it("Should render correctly on small screens", () => {
			const { container } = render(
				<TabControls currentWeatherDisabled={false} largeScreen={false} onTabChange={voidCallback} currentTab={0} />,
			);

			expect(container.firstChild).toMatchSnapshot();
		});

		it("Should render correctly with disabled current weather tab", () => {
			const { container } = render(
				<TabControls currentWeatherDisabled largeScreen onTabChange={voidCallback} currentTab={0} />,
			);

			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
