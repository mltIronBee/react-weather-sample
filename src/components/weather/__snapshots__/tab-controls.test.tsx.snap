// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab controls component Rendering Should render correctly on large screen 1`] = `
<div
  class="MuiTabs-root"
>
  <div
    class="MuiTabs-scroller MuiTabs-fixed"
    style="overflow: hidden;"
  >
    <div
      aria-label="tabs control to pick between week forecast and current weather"
      class="MuiTabs-flexContainer MuiTabs-centered"
      role="tablist"
    >
      <button
        aria-controls="forecast-panel"
        aria-label="forecast tab"
        aria-selected="true"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-selected"
        id="forecast-tab"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTab-wrapper"
        >
          Forecast
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-controls="current-weather-panel"
        aria-label="current weather tab"
        aria-selected="false"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
        id="current-weather-tab"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiTab-wrapper"
        >
          Current Weather
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <span
      class="PrivateTabIndicator-root PrivateTabIndicator-colorPrimary MuiTabs-indicator"
      style="left: 0px; width: 0px;"
    />
  </div>
</div>
`;

exports[`Tab controls component Rendering Should render correctly on small screens 1`] = `
<div
  class="MuiTabs-root"
>
  <div
    class="MuiTabs-scroller MuiTabs-fixed"
    style="overflow: hidden;"
  >
    <div
      aria-label="tabs control to pick between week forecast and current weather"
      class="MuiTabs-flexContainer"
      role="tablist"
    >
      <button
        aria-controls="forecast-panel"
        aria-label="forecast tab"
        aria-selected="true"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-selected MuiTab-fullWidth"
        id="forecast-tab"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTab-wrapper"
        >
          Forecast
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-controls="current-weather-panel"
        aria-label="current weather tab"
        aria-selected="false"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-fullWidth"
        id="current-weather-tab"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiTab-wrapper"
        >
          Current Weather
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <span
      class="PrivateTabIndicator-root PrivateTabIndicator-colorPrimary MuiTabs-indicator"
      style="left: 0px; width: 0px;"
    />
  </div>
</div>
`;

exports[`Tab controls component Rendering Should render correctly with disabled current weather tab 1`] = `
<div
  class="MuiTabs-root"
>
  <div
    class="MuiTabs-scroller MuiTabs-fixed"
    style="overflow: hidden;"
  >
    <div
      aria-label="tabs control to pick between week forecast and current weather"
      class="MuiTabs-flexContainer MuiTabs-centered"
      role="tablist"
    >
      <button
        aria-controls="forecast-panel"
        aria-label="forecast tab"
        aria-selected="true"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-selected"
        id="forecast-tab"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiTab-wrapper"
        >
          Forecast
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        aria-controls="current-weather-panel"
        aria-label="current weather tab"
        aria-selected="false"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-disabled MuiButtonBase-disabled"
        disabled=""
        id="current-weather-tab"
        role="tab"
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiTab-wrapper"
        >
          Current Weather
        </span>
      </button>
    </div>
    <span
      class="PrivateTabIndicator-root PrivateTabIndicator-colorPrimary MuiTabs-indicator"
      style="left: 0px; width: 0px;"
    />
  </div>
</div>
`;
